neqo-transport/src/packet.rs

```diff
@@ -24,745 +27,931 @@ const PACKET_TYPE_RETRY: u8 = 0x03;
 
 const PACKET_BIT_LONG: u8 = 0x80;
 const PACKET_BIT_SHORT: u8 = 0x00;
+const PACKET_BIT_KEY_PHASE: u8 = 0x04;
 const PACKET_BIT_FIXED_QUIC: u8 = 0x40;
 
+const PACKET_HP_MASK_LONG: u8 = 0x0f;
+const PACKET_HP_MASK_SHORT: u8 = 0x1f;
+
 const SAMPLE_SIZE: usize = 16;
+const SAMPLE_OFFSET: usize = 4;
 
-const AUTH_TAG_LEN: usize = 16;
+pub type PacketNumber = u64;
 
-#[derive(Debug, PartialEq)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
 pub enum PacketType {
-    Short(bool), // Includes whether the key phase is set
-    ZeroRTT,
+    VersionNegotiation,
+    Initial,
     Handshake,
-    VN(Vec<Version>), // List of versions
-    Initial(Vec<u8>), // Token
-    Retry { odcid: ConnectionId, token: Vec<u8> },
+    ZeroRtt,
+    Retry,
+    Short,
+    OtherVersion,
 }
 
 impl PacketType {
     #[must_use]
-    fn code(&self) -> u8 {
+    fn code(self) -> u8 {
         match self {
-            Self::Initial(..) => PACKET_TYPE_INITIAL,
-            Self::ZeroRTT => PACKET_TYPE_0RTT,
+            Self::Initial => PACKET_TYPE_INITIAL,
+            Self::ZeroRtt => PACKET_TYPE_0RTT,
             Self::Handshake => PACKET_TYPE_HANDSHAKE,
-            Self::Retry { .. } => PACKET_TYPE_RETRY,
+            Self::Retry => PACKET_TYPE_RETRY,
             _ => panic!("shouldn't be here"),
         }
     }
-
-    #[must_use]
-    pub fn space(&self) -> PNSpace {
-        match self {
-            Self::Short(_) | Self::ZeroRTT => PNSpace::ApplicationData,
-            Self::Handshake => PNSpace::Handshake,
-            Self::Initial(_) => PNSpace::Initial,
-            _ => panic!("don't ask for the space when there isn't one"),
-        }
-    }
-
-    #[must_use]
-    pub fn key_phase(&self) -> bool {
-        matches!(self, Self::Short(true))
-    }
 }
 
-impl Default for PacketType {
-    fn default() -> Self {
-        Self::Short(false)
+/// The AEAD used for Retry is fixed, so use this.
+fn make_retry_aead() -> Aead {
+    #[cfg(debug_assertions)]
+    ::neqo_crypto::assert_initialized();
+
+    let secret = hkdf::import_key(
+        TLS_VERSION_1_3,
+        TLS_AES_128_GCM_SHA256,
+        &[
+            0x65, 0x6e, 0x61, 0xe3, 0x36, 0xae, 0x94, 0x17, 0xf7, 0xf0, 0xed, 0xd8, 0xd7, 0x8d,
+            0x46, 0x1e, 0x2a, 0xa7, 0x08, 0x4a, 0xba, 0x7a, 0x14, 0xc1, 0xe9, 0xf7, 0x26, 0xd5,
+            0x57, 0x09, 0x16, 0x9a,
+        ],
+    )
+    .unwrap();
+    Aead::new(TLS_VERSION_1_3, TLS_AES_128_GCM_SHA256, &secret, "quic ").unwrap()
+}
```

Comment:
Would crypto.rs be a better home for this?

Also, just from looking at quic-tls  "5.8. Retry Packet Integrity " as a non-crypto expert, I see where the 0x65.. is coming from, but the label, why is it "quic " here?

Reply:
This is what we do for all of the AEAD construction, it's the extra piece that distinguishes the QUIC keys from regular TLS 1.3 keys.